rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,491 @@
+latencies that when not serviced, lead to other processing.  This is worth
+mentioning because it will be important to compare a deployment to itself, and
+there may be limited value in comparing across deployments.
+
+## Capella Service Level Metrics
+
+The counters described above give a good picture of actual impact on a deployed
+app, but they do not give a sense of whether or not a given deployment is
+meeting an expected service level.  Said another way, a Couchbase Capella
+Cluster deployment could be operational, but very slow during a rebalance
+leading to user discontent, but never turn into an error.
+
+Service levels are specific to the user's targets.  While it's impossible to
+properly measure whether or not an end-user's app is meeting service level
+targets, we can establish a predefined histogram that would allow Couchbase to
+simply and inexpensively understand if the service level has likely changed.
+
+It's proposed that we establish a set of standardized histograms, specific to
+services and to the service level expected with the service using those
+parameters.  Counters for these histograms can be sent to Couchbase and both
+stored individually and aggregated. See histogram buckets in "Histograms" table
+below.
+
+As an aggregated set of metrics, it would give Couchbase a rough view over the
+service level the user is experiencing.  Deviations in service level can be dug
+into and, in particular, we can check for correlation to Couchbase Capella
+management activities like upgrades.
+
+The Indexing Service and Eventing Services are not included as they are not
+network-available services for applications to make requests from.
+
+As with the Service Experience Counters, it may be incorrect to try to
+compare between deployments.  For example, the query complexity may differ
+significantly or the data size and use of features like compression may vary
+significantly between two applications.
+
+# API and Implementation Details
+
+The decision to reuse the existing Meter implementation is considered an SDK
+implementation detail, and this RFC does not specify the API for collecting
+metrics from the library.
+
+One possible approach is to introduce a "compound meter" that aggregates an
+existing meter (which could be overridden by the user) and the application
+service-level telemetry meter described in this RFC. This compound meter would
+collect counters and fixed-bucket latency histograms. The application
+service-level meter would continue to function even if the user provides a
+custom meter implementation.
+
+In addition to latencies, which would be used for reporting histograms, the
+default Meter should also maintain the following counters:
+
+| Metric | Description |
+| :---- | :---- |
+| sdk\_*{service}*\_r\_utimedout | Unambiguous timeout |
+| sdk\_*{service}*\_r\_atimedout | Ambiguous timeout |
+| sdk\_*{service}*\_r\_canceled | Canceled |
+| sdk\_*{service}*\_r\_total | Total number of operations |
```

Comment:
How does the `total` value differ from the histogram `count`?

Reply:
@dnault total is total number of the operations, count is number of the data points in histogram. Histograms do not account for errors.

Reply:
Updated to say it explicitly in the spec.